.github/workflows/ci-cd.yml

```diff
@@ -32,10 +35,34 @@ jobs:
           echo "MSYS2 packages: ${msys2_packages}"
           echo "::set-output name=msys2_packages::${msys2_packages}"
 
+      - name: Detect changed files
+        if: ${{ github.event_name == 'pull_request' }}
+        run: git diff --name-only "${{ github.base_ref }}..HEAD" > _changed_files
+
+      - name: Detect changed files
+        if: ${{ github.event_name == 'push' }}
+        run: git diff --name-only "${{ github.event.before }}..HEAD" > _changed_files
+
+      - name: Detect changed packages
+        id: changes
+        run: |
+          if grep '^packages/(server|front|build-tools)' > /dev/null < _changed_files; then
```

Comment:
Nice/portable grep. I don't think it matters as we are controlling the environment the grep runs it.

You could be able to ditch `<` (the stdin redirection) as grep's second positional parameter is a file.